hnn_core/network.py

```diff
@@ -947,29 +947,47 @@ def _all_to_all_connect(self, src_cell, target_cell,
             all target_type cells.
         """
         src_gids = self.gid_ranges[_long_name(src_cell)]
-        target_gids = self.gid_ranges[_long_name(target_cell)]
+        target_range = self.gid_ranges[_long_name(target_cell)]
 
         src_start = src_gids[0]  # Necessary for unique feeds
-        for target_gid in target_gids:
-            if unique:
-                src_gids = [target_gid + src_start]
-            for src_gid in src_gids:
-                if not allow_autapses and src_gid == target_gid:
-                    continue
-                self.add_connection(
-                    src_gid, target_gid, loc, receptor,
-                    nc_dict['A_weight'], nc_dict['A_delay'], nc_dict['lamtha'])
 
-    def add_connection(self, src_gid, target_gid, loc, receptor,
+        if unique:
+            src_gids = np.array(target_range) + src_start
+            src_gids = src_gids.astype(int).tolist()
+            target_gids = [[target_gid] for target_gid in target_range]
+        else:
+            target_gids = list()
+            for src_gid in src_gids:
+                target_src_pair = list()
+                for target_gid in target_range:
+                    if not allow_autapses and src_gid == target_gid:
+                        continue
+                    target_src_pair.append(target_gid)
+                target_gids.append(target_src_pair)
+
+        self.add_connection(
+            src_gids, target_gids, loc, receptor,
+            nc_dict['A_weight'], nc_dict['A_delay'], nc_dict['lamtha'])
+
+    def add_connection(self, src_gids, target_gids, loc, receptor,
                        weight, delay, lamtha):
         """Appends connections to connectivity list
 
         Parameters
         ----------
-        src_gid : int
-            Integer identifying source cell ID.
-        target_gid : list | range | int
-            Integer or list of integers identifying target cell ID.
+        src_gids : str | int | range | list of int
+            Identifier for src cells. Passing str arguments
+            ('L2_pyramidal', 'L2_basket', 'L5_pyramidal', 'L5_basket') is
+            equivalent to passing a list of gids for the relvant cell type.
+            src-target connections depend on the input type of target_gids.
+        target_gids : str | int | range | list of int | list of list of int
+            Identifer for targets of src cell. Passing str arguments
+            ('L2_pyramidal', 'L2_basket', 'L5_pyramidal', 'L5_basket') is
+            equivalent to passing a list of gids for the relvant cell type.
+            Inputs of type (str, int, range, and list of int) connect every
+            src_gid to an identical set of targets.
+            Targets can be uniquely specified for each src_gid by passing a
+            list of lists (must contain one element for each src_gid).
```

Comment:
this is overloading the `target_gids` param to pass the `connectivity`. I think it's a little confusing semantically. If a user wants this level of control. I think they can write a simple for loop:

```py
for src_gid, target_gids in conns:
    net.add_connections(src_gid, target_gids, ...)
```

Reply:
Not against this change, but it will increase the size of net.connectivity for the default network (when allow_autapses=False). Might be negligible, but may become problematic for larger networks.

Reply:
Let's worry about that later. I'm trying to push back a little against giving users too fine-level of control through the API. API should be high-level but advanced users should be able to do more sophisticated stuff by using some small for loop or function.

Reply:
Just implemented the change. It's actually not a huge size increase in `net.connectivity` 5MB, but it does make `net.connectivity` a bit less clear when inspecting it's contents. Specifically, the previous version had 16 elements for each unique connection class of the cortical column. Taking away overloading expands this to >1000 elements. 

Perhaps we can simply not advertise the overloaded functionality? The default network instantiation uses `net.add_connection`, and the code is a great deal simpler

Reply:
okay I need to play around with this a bit. Do you have any more local changes? I won't get to it tomorrow. Maybe on Monday ...

Reply:
I wouldn't advertise complicated functionality. If someone asks, you can show them how to do it. It's what we have done in MNE in the past. But the `net.connectivity` being harder to inspect (in the default case?) is indeed a little worrisome.

Reply:
To be clearer, [a24af9d](https://github.com/jonescompneurolab/hnn-core/pull/313/commits/a24af9db597c8eb4df1c40bdc239ed91e71e01a9) is form I'd like to keep. `net.connectivity` simply has 16 elements in the list. Overloading `net.add_connection` allows target_gids to vary across srcs within the same element of the list.

Removing overloading in [6257db8](https://github.com/jonescompneurolab/hnn-core/pull/313/commits/6257db861ac848f69691a874572b1cf517ebe547) forces `net._all_to_all_connect` to call `net.add_connection` multiple times in a for loop which makes the default connectivity more confusing upon inspection. 

I'm arguing that we allow overloading to maintain a simpler default connectivity object, but refrain from advertising it in the docs or examples. It will be extremely difficult for a user to accidentally invoke this overloaded functionality given how precise the format must be.

Reply:
okay after some thought and playing around, I agree that the overloading keeps the connectivity simple. But let's not advertise it in docs and examples.

however, I think there is a bit more to be done to make the structure more transparent. Can you add back `src_type` and `target_type` in the dictionary. It may increase the size a bit but it's the price we pay for transparent objects. And then, if you can make a `_Connectivity(dict)` with a `repr` that hides the `gid_pairs` it will make inspecting connectivity much easier.

Reply:
Btw, instead of `gid_pairs`, if you put `src_gids` and `target_gids` in `net.connectivity`, wouldn't you be able to retain the simple connectivity structure even with `allow_autapses=False`? Then in `network_builder.py` you construct the `gid_pairs` by doing `itertools.product(src_gids, target_gids)`

Reply:
That sounds good to me! Are you wanting to enforce the same `src_type` and `target_type` for each connection? It is technically possible to have a mix of cell classes at the moment, it just doesn't show up in the default instantiation.